src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+
+    def buffer(self, amount):
+
+        buffered_polygons = self.polygons.buffer(amount)
+
+        return PolygonalRegion(polygon=buffered_polygons, name=self.name, z=self.z)
+
+    @property
+    def dimensionality(self):
+        return 2
+
+    @cached_property
+    def size(self):
+        return self.polygons.area
+
+    @property
+    def area(self):
+        return self.size
+
+    def show(self, plt, style='r-', **kwargs):
+        plotPolygon(self.polygons, plt, style=style, **kwargs)
+
+    def __repr__(self):
+        return f'PolygonalRegion({self.polygons!r})'
+
+    def __eq__(self, other):
+        if type(other) is not PolygonalRegion:
+            return NotImplemented
+        return (other.polygons == self.polygons
+                and other.orientation == self.orientation)
+
+    @cached
+    def __hash__(self):
+        return hash((self.polygons, self.orientation))
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state.pop('_cached_prepared', None)     # prepared geometries are not picklable
+        return state
+
+class CircularRegion(PolygonalRegion):
+    """A circular region with a possibly-random center and radius.
+
+    Args:
+        center (`Vector`): center of the disc.
+        radius (float): radius of the disc.
+        resolution (int; optional): number of vertices to use when approximating this region as a
+            polygon.
+        name (str; optional): name for debugging.
+    """
+    def __init__(self, center, radius, resolution=32, name=None):
+        self.center = toVector(center, "center of CircularRegion not a vector")
+        self.radius = toScalar(radius, "radius of CircularRegion not a scalar")
+        self.circumcircle = (self.center, self.radius)
+        self.resolution = resolution
+
+        deps = [self.center, self.radius]
+
+        if any(needsSampling(dep) for dep in deps):
+            # Center and radius aren't fixed, so we'll wait to pass a polygon
+            super().__init__(polygon=None, name=name, additionalDeps=deps)
+            return
+
+        ctr = shapely.geometry.Point(self.center)
+        polygons = MultiPolygon([ctr.buffer(self.radius, resolution=self.resolution)])
+        super().__init__(polygon=polygons, z=self.center.z, name=name, additionalDeps=deps)
+
+    def sampleGiven(self, value):
+        return CircularRegion(value[self.center], value[self.radius],
+                              name=self.name, resolution=self.resolution)
+
+    def evaluateInner(self, context):
+        center = valueInContext(self.center, context)
+        radius = valueInContext(self.radius, context)
+        return CircularRegion(center, radius, name=self.name, resolution=self.resolution)
+
+    def intersects(self, other, triedReversed=False):
+        if isinstance(other, CircularRegion):
+            return self.center.distanceTo(other.center) <= self.radius + other.radius
+        
+        return super().intersects(other)
```

Comment:
Missing `triedReversed`.

There were enough of these problems that I think we'd better write an omnibus test that creates 1 region of each type and tries all possible calls to `intersects`, `intersect`, `union`, etc. The combinations that are expected to fail should be explicitly listed and required to raise a specific `NotImplementedError` or other appropriate exception rather than failing in some other way.

Reply:
Good idea. Fixed the triedReversed and put the test on the list to add.